src/verilog/verilog_elaborate_constants.cpp

```diff
@@ -1,144 +1,271 @@
 /*******************************************************************\
 
-Module: Verilog Elaboration Time Constants
+Module: Verilog Elaboration
 
 Author: Daniel Kroening, kroening@kroening.com
 
 \*******************************************************************/
 
+#include <util/arith_tools.h>
+#include <util/bitvector_types.h>
+#include <util/mathematical_types.h>
+
 #include "verilog_typecheck.h"
+#include "verilog_types.h"
 
-/*******************************************************************\
+void verilog_typecheckt::collect_symbols(
+  const typet &type,
+  const verilog_parameter_declt::declaratort &declarator)
+{
+  // A localparam or parameter declarator.
+  auto base_name = declarator.identifier();
 
-Function: verilog_typecheckt::elaborate_constants
+  auto full_identifier =
+    id2string(module_identifier) + "." + id2string(base_name);
 
-  Inputs:
+  // If there's no type, parameters take the type of the
+  // value. We signal this using the special type "derive_from_value".
 
- Outputs:
+  auto symbol_type = type_with_subtypet(
+    ID_to_be_elaborated, type.is_nil() ? typet("derive_from_value") : type);
 
- Purpose:
+  symbolt symbol{full_identifier, symbol_type, mode};
 
-\*******************************************************************/
+  symbol.module = module_identifier;
+  symbol.base_name = base_name;
+  symbol.pretty_name = strip_verilog_prefix(symbol.name);
+  symbol.is_macro = true;
+  symbol.value = declarator.value();
+  symbol.location = declarator.source_location();
+
+  add_symbol(std::move(symbol));
+}
 
-void verilog_typecheckt::elaborate_constants()
+void verilog_typecheckt::collect_symbols(const typet &type)
 {
-  // This refers to "elaboration-time constants" as defined
-  // in System Verilog 1800-2017, and includes
-  // * parameters (including parameter ports)
-  // * localparam
-  // * specparam
-  // * enum constants
-  //
-  // These may depend on each other, in any order.
-  // We traverse these dependencies recursively.
-  // First collect all constant identifiers into the symbol table,
-  // with type "to_be_elaborated".
+  // These types are not yet converted.
+  if(type.id() == ID_verilog_enum)
+  {
+    // These have a body, with enum constants, and an optional base type.
+    const auto &enum_type = to_verilog_enum_type(type);
+
+    if(enum_type.has_base_type())
+      collect_symbols(enum_type.base_type());
+
+    /*
+    // In contrast to C, there is no tag in the syntax, but our parser
+    // generates one. We use that as type for the enum names.
+    const verilog_enum_tag_typet enum_tag_type(enum_type.identifier());
+    */
```

Comment:
What to make of this commented-out block?

Reply:
It's one of the KNOWNBUG tests -- enums in Verilog are meant to be strictly typed, and I am wondering whether to achieve that with a tag. But there's likely no need, given that the enum names must be unique.